src/main/java/com/microsoft/azure/relay/HybridConnectionListener.java

```diff
@@ -276,46 +276,42 @@ public HybridConnectionListener(String connectionString, String path) throws URI
 	 * @return A CompletableFuture which completes when all connections are
 	 *         disconnected with the cloud service
 	 */
-	@SuppressWarnings("unchecked")
 	public CompletableFuture<Void> closeAsync(Duration timeout) {
-		return CompletableFutureUtil.timedSupplyAsync(timeout, () -> {
-			synchronized (this.thisLock) {
-				if (this.closeCalled) {
-					return CompletableFuture.completedFuture(null);
-				}
+		TimeoutHelper timeoutHelper = new TimeoutHelper(timeout);
+		CompletableFuture<?>[] closeTasks;
+		synchronized (this.thisLock) {
+			if (this.closeCalled) {
+				return CompletableFuture.completedFuture(null);
+			}
 
-				RelayLogger.logEvent("closing", this);
-				this.closeCalled = true;
+			RelayLogger.logEvent("closing", this);
+			this.closeCalled = true;
 
-				// If the input queue is empty this completes all pending waiters with null and
-				// prevents any new items being added to the input queue.
-				this.connectionInputQueue.shutdown();
+			// If the input queue is empty this completes all pending waiters with null and
+			// prevents any new items being added to the input queue.
+			this.connectionInputQueue.shutdown();
 
-				// Close any unaccepted rendezvous. DequeueAsync won't block since we've called
-				// connectionInputQueue.Shutdown().
-				CompletableFuture<?>[] closeTasks = new CompletableFuture<?>[this.connectionInputQueue.getPendingCount()];
-				for (int i = 0; i < this.connectionInputQueue.getPendingCount(); i++) {
-					closeTasks[i] = this.connectionInputQueue.dequeueAsync().thenAccept(connection -> {
-						connection.closeAsync(
-								new CloseReason(CloseReason.CloseCodes.NORMAL_CLOSURE, "Client closing the socket normally"));
-					});
-				}
-				return closeTasks;
+			// Close any unaccepted rendezvous. DequeueAsync won't block since we've called
+			// connectionInputQueue.Shutdown().
+			closeTasks = new CompletableFuture<?>[this.connectionInputQueue.getPendingCount()];
+			for (int i = 0; i < this.connectionInputQueue.getPendingCount(); i++) {
+				closeTasks[i] = this.connectionInputQueue.dequeueAsync(timeoutHelper.remainingTime()).thenAccept(connection -> {
+					connection.closeAsync(new CloseReason(CloseReason.CloseCodes.NORMAL_CLOSURE, "Client closing the socket normally"));
+				});
 			}
-		}, EXECUTOR)
-		.thenCompose(closeTasks -> {
-			return CompletableFuture.allOf((CompletableFuture<Void>[]) closeTasks).thenRun(() -> {
+		}
+		
+		return CompletableFuture.allOf(closeTasks)
+			.thenCompose(voidResult -> this.controlConnection.closeAsync(timeoutHelper.remainingTime()))
+			.whenComplete((voidResult, ex) -> {
 				this.isOnline = false;
 				RelayLogger.logEvent("closed", this);
```

Comment:
>RelayLogger.logEvent("closed", this); [](start = 4, length = 37)

maybe this log should be at the very end?

Reply:
I can move it after connectionInputQueue.dispose(), or is there somewhere else you'd like it to be?

---
In reply to: [258612564](https://github.com/Azure/azure-relay-java/pull/15#discussion_r258612564) [](ancestors = 258612564)

Reply:
That works

---
In reply to: [258613118](https://github.com/Azure/azure-relay-java/pull/15#discussion_r258613118) [](ancestors = 258613118,258612564)

Reply:
Fixed in Iteration 2.

---
In reply to: [258613625](https://github.com/Azure/azure-relay-java/pull/15#discussion_r258613625) [](ancestors = 258613625,258613118,258612564)